core/src/main/scala/com/eed3si9n/jarjarabrams/Zip.scala

```diff
@@ -1,192 +1,134 @@
-      output putNextEntry makeFileEntry(file, name)
-      transfer(file, output)
-      output.closeEntry()
-    }
-
-    // Calculate directories and add them to the generated Zip
-    allDirectoryPaths(files).foreach(addDirectoryEntry(_))
-
-    // Add all files to the generated Zip
-    files.foreach { case (file, name) => addFileEntry(file, name) }
-  }
-
-  private def normalizeToSlash(name: String) = {
-    val sep = File.separatorChar
-    if (sep == '/') name else name.replace(sep, '/')
-  }
-
-  // map a path a/b/c to List("a", "b")
-  private def relativeComponents(path: String): List[String] =
-    path.split("/").toList.dropRight(1)
-
-  // map components List("a", "b", "c") to List("a/b/c/", "a/b/", "a/", "")
-  private def directories(path: List[String]): List[String] =
-    path.foldLeft(List(""))((e, l) => (e.head + l + "/") :: e)
-
-  // map a path a/b/c to List("a/b/", "a/")
-  private def directoryPaths(path: String): List[String] =
-    directories(relativeComponents(path)).filter(_.length > 1)
-
-  // produce a sorted list of all the subdirectories of all provided files
-  private def allDirectoryPaths(files: Iterable[(Path, String)]) =
-    TreeSet[String]() ++ (files.flatMap { case (_, name) => directoryPaths(name) })
-
-  private def withZipOutput(file: Path, manifest: Option[Manifest], time: Long)(
-      f: ZipOutputStream => Unit
-  ) = {
-    Using.fileOutputStream(false)(file) { fileOut =>
-      val (zipOut, _) =
-        manifest match {
-          case Some(mf) =>
-            import Attributes.Name.MANIFEST_VERSION
-            val main = mf.getMainAttributes
-            if (!main.containsKey(MANIFEST_VERSION))
-              main.put(MANIFEST_VERSION, "1.0")
-            val os = new JarOutputStream(fileOut)
-            val e = new ZipEntry(JarFile.MANIFEST_NAME)
-            e.setTime(time)
-            os.putNextEntry(e)
-            mf.write(new BufferedOutputStream(os))
-            os.closeEntry()
-            (os, "jar")
-          case None => (new ZipOutputStream(fileOut), "zip")
-        }
-      try {
-        f(zipOut)
-      } finally {
-        zipOut.close
+  // ZIP timestamps have a resolution of 2 seconds.
+  // see http://www.info-zip.org/FAQ.html#limits
+  final private val minimumTimestampIncrement = 2000L
+  // 1980-01-03
+  // Technically it should be 1980-01-01, but Java checks against
+  // the year segment in localtime and ends up capturing timezone,
+  // so to be safe we should consider Jan 3 to be the minimum.
+  //
+  //   $ unzip -l example/byte-buddy-agent.jar | head
+  // Archive:  example/byte-buddy-agent.jar
+  //   Length      Date    Time    Name
+  // ---------  ---------- -----   ----
+  //         0  00-00-1980 04:08   META-INF/
+  //       988  00-00-1980 04:08   META-INF/MANIFEST.MF
+  private final val minimumTimestamp = 315705600L
+
+  def list(inputJar: Path): List[(String, Long)] =
+    Using.jarFile(inputJar) { in =>
+      in.entries.asScala.toList.map { entry =>
```

Comment:
Nit: I would do

`.asScala.iterator.map(...).toList`

To avoid allocating an intermediate List.

Reply:
Addressed in https://github.com/eed3si9n/jarjar-abrams/pull/39/commits/d41bfc7e682966948cea777d4cbe8e9db4ca1b65